# Systematic Biology Lab Repository
I will post the instructions code necessary for the computational exercises that we will do in the class.

## Alignments

The first module is on multiple sequence alignments and it is divided in two parts:

[Alignments Part 1: MAFFT Online and Mesquite](https://cjpardodelahoz.github.io/BIO556L/alignments_online)

[Alignments Part 2: MAFFT in the Command Line on the DCC](https://cjpardodelahoz.github.io/BIO556L/alignments_dcc)

## Models of Evolution and Tree Inference

We are then going to look at different ways of selecting models of sequence evolution and how to use them in likelihood and bayesin tree inference. This is divided in two parts:

[Choosing Models of Sequence Evolution](https://cjpardodelahoz.github.io/BIO556L/models)

[Bayesian Phylogenetic Inference](https://cjpardodelahoz.github.io/BIO556L/bayes)


## Sources of Phylogenetic Incongruence

Now we will dive into different biological sources of phylogenetic incongruence and the methods that have been developed to account for it in inference.

[Species Trees with SVDQuartets (Dave Swofford guest tutorial)](https://phylosolutions.com/tutorials/wh2022-svdq-astral/species-trees-tutorial.html)

[Species Trees and Gene Tree Heterogeneity due to ILS](https://cjpardodelahoz.github.io/BIO556L/conflict)

[Phylogenetic Network Inference with SNaQ](https://cjpardodelahoz.github.io/BIO556L/snaq)

## Species delimitation

[Joint Species Tree Estimation and Species Delimitation with BPP](https://cjpardodelahoz.github.io/BIO556L/bpp)

## Divergence Time Estimation

[BEAST Tutorial](https://beast2-dev.github.io/beast-docs/beast2/DivergenceDating/DivergenceDatingTutorial.html)

[Download Beast version 2.3.2](https://github.com/CompEvol/beast2/releases?page=3)
